lang/ast/src/exp/mod.rs

```diff
@@ -120,17 +120,23 @@ impl Shift for Exp {
 }
 
 impl Occurs for Exp {
-    fn occurs(&self, ctx: &mut LevelCtx, lvl: Lvl) -> bool {
+    fn occurs<F>(&self, ctx: &mut LevelCtx, f: &F) -> bool
+    where
+        F: Fn(&LevelCtx, &Exp) -> bool,
+    {
+        if f(ctx, self) {
+            return true;
+        }
         match self {
-            Exp::Variable(e) => e.occurs(ctx, lvl),
-            Exp::TypCtor(e) => e.occurs(ctx, lvl),
-            Exp::Call(e) => e.occurs(ctx, lvl),
-            Exp::DotCall(e) => e.occurs(ctx, lvl),
-            Exp::Anno(e) => e.occurs(ctx, lvl),
-            Exp::TypeUniv(e) => e.occurs(ctx, lvl),
-            Exp::LocalMatch(e) => e.occurs(ctx, lvl),
-            Exp::LocalComatch(e) => e.occurs(ctx, lvl),
-            Exp::Hole(e) => e.occurs(ctx, lvl),
+            Exp::Variable(_) => false,
```

Comment:
Maybe add a comment here: Variables have no subexpressions, therefore the check above is sufficient. Similarly for TypeUniv.